### PR TITLE
Simulate: Properly handle failing clear state programs

### PR DESCRIPTION
**Pull request by jasonpaulos**
## Problem

Currently the simulate endpoint does not properly handle failing clear state programs. Unlike every other type of program execution, if a clear state program rejects or errors, the transaction group continues and can still succeed. Any stateful operations the clear state program performed before it failed are rolled back.

Simulate incorrectly handled failing clear state programs in the following ways:
1. Reporting the wrong ApplyData for a failed clear state txn (it should be blank, but that's not what simulate reports)
2. If there's a later error, it uses the location of the earlier clear state error
3. For app state traces, it doesn't communicate that clear state operations should be rolled back

## Solution

Implementation fixes for items 1 & 2 have been added in this PR.

Additionally, for item 3, the fields `clear-state-rollback` and `clear-state-rollback-error` have been added to the `SimulationTransactionExecTrace` object. If `clear-state-rollback` is true, this indicates to trace consumers that they must discard all state changes from the clear state trace after it finishes.

## Test Plan

Unit tests added

### CODE REVIEW

**Review by jannotti — APPROVED** [review]

File: data/transactions/logic/mocktracer/tracer.go [hunk 1]
```diff
@@ -112,8 +115,8 @@ func AfterTxn(txnType protocol.TxType, ad transactions.ApplyData, hasError bool)
 }
 
 // AfterProgram creates a new Event with the type AfterProgramEvent
-func AfterProgram(mode logic.RunMode, hasError bool) Event {
-	return Event{Type: AfterProgramEvent, LogicEvalMode: mode, HasError: hasError}
+func AfterProgram(mode logic.RunMode, pass bool, hasError bool) Event {
```
Comment (algorandskiy) [inline data/transactions/logic/mocktracer/tracer.go:118]: suggestion: two boolean flags in a row do not look great, maybe invent a small `programStatus` type with values "OK" (pass=T, err=F), "REJ" (pass=F, err=F), "ERR" (pass=F, err=T) - assuming pass=T, err=T is impossible combo.

File: ledger/simulation/tracer.go [hunk 2]
```diff
@@ -185,12 +185,14 @@ func (tracer *evalTracer) AfterTxnGroup(ep *logic.EvalParams, deltas *ledgercore
...
 }
 
-func (tracer *evalTracer) saveApplyData(applyData transactions.ApplyData) {
+func (tracer *evalTracer) saveApplyData(applyData transactions.ApplyData, omitEvalDelta bool) {
 	applyDataOfCurrentTxn := tracer.mustGetApplyDataAtPath(tracer.absolutePath())
 	// Copy everything except the EvalDelta, since that has been kept up-to-date after every op
 	evalDelta := applyDataOfCurrentTxn.EvalDelta
 	*applyDataOfCurrentTxn = applyData
-	applyDataOfCurrentTxn.EvalDelta = evalDelta
+	if omitEvalDelta {
```
Comment (algorandskiy) [inline ledger/simulation/tracer.go:193]: "omit" means skip, right? why is it assigned in this case? Like the condition should be reversed?

Comment (jasonpaulos) [reply ledger/simulation/tracer.go:193]: Good question. The `evalDelta` variable is the old data, so if `omitEvalDelta` is true, we keep the old data in `applyDataOfCurrentTxn`. If false then we take the new data from `applyData`.

Comment (jasonpaulos) [reply ledger/simulation/tracer.go:192]: Added comment in 8fd295f56

File: data/transactions/logic/mocktracer/tracer.go (hunk 1 above)
Comment (jasonpaulos) [reply data/transactions/logic/mocktracer/tracer.go]: I like this, it's easier to understand. Made change in 8fd295f56

**Review by algorandskiy — APPROVED** [review]

**Review by jannotti — APPROVED** [review]

**Pull request merged**